core/src/mast/serialization/mod.rs

```diff
@@ -90,27 +94,13 @@ impl Serializable for MastForest {
         target.write_bytes(MAGIC);
         target.write_bytes(&VERSION);
 
-        // decorator & node counts
-        target.write_usize(self.decorators.len());
+        // node counts
```

Comment:
I'm not sure in which context we intend to actually drop the decorator data from the serialized representation, but I'm wondering: How do we actually do that? We only have a `Vec<u8>` so wouldn't we need the length of the decorator data at some fixed offset to read it and then be able to drop the appropriate amount of bytes from the end? Since we can't retroactively write at an offset at the beginning of the section, perhaps we could add this offset at the very end of the layout.

```
<MAGIC>
<VERSION>
...
<DECORATOR_DATA_OFFSET_FROM_BACK: u64>
```

This way we could read the `u64` (or whatever makes sense, it just needs a fixed size) from the back of the `Vec<u8>`.
Then we drop the the decorator data:

```rust
let mut serialized = vec![...];
let offset = serialized.last().unwrap();
serialized.truncate(serialized.len() - (offset + core::mem::size_of::<u64>()));
```

Reply:
@PhilippGackstatter Not sure what the best mechanism would be, but your proposal seems reasonable.

Reply:
> I'm not sure in which context we intend to actually drop the decorator data from the serialized representation, but I'm wondering: How do we actually do that?

I think the general idea is that producers, e.g. the compiler, will emit a fully-fleshed out package containing all the things that could be useful for interacting with that package (e.g. debug info, documentation, sources in some caes, etc.), and the MAST contained in the package would have all of the decorators (naturally required to support debugging, etc.). There are contexts where all of that stuff is simply dead weight though, or you want to optimize for binary size, in which case you can strip all of that, resulting in a minimal package that only contains exactly what is needed to execute what it contains. One such potential use case, would be publishing packages on-chain, which would only be feasible if they are as small as possible.

My assumption was that we stored decorators in a separate section, whose offset and size is recorded in the header of the binary format. Decorator references are simply relative offsets to the decorator data stored in the decorator section (i.e. the offset is relative to the base offset of the decorator section). Thus, truncating the decorators conceptually involves splitting the binary at the boundaries of the decorator section, then stitching the parts we're keeping back together, and setting the decorator section size and offset to `0`. Decorator references would remain, but would be ignored since there is no decorator section to resolve them with.

Stripping both the decorator data _and_ the decorator references, probably requires reconstructing the MAST from scratch, since I don't think we have any means by which to keep both references and the data separate from everything else. One slightly less drastic approach might be to visit all the decorator references and zero them out, so that compressing the binary format essentially erases any overhead associated with leaving them behind.

I think fundamentally though, the goal is to provide a fast and efficient means of stripping optional bits from the MAST, without rebuilding it from scratch. Optional sections would be kept separate as described above, so that you don't have to really know anything about what is stored in them to remove them.

That's my take anyway!

Reply:
Thank you for the explanation on where this is useful!

> Stripping both the decorator data and the decorator references, probably requires reconstructing the MAST from scratch, since I don't think we have any means by which to keep both references and the data separate from everything else.

Maybe I'm missing something but I think that is possible. Currently the before-enter and after-exit decorator data is already completely separate. It's a `Vec<(MastNodeId, Vec<DecoratorId>)>` where for each mast node id we add the vector of decorator ids. If that is missing, we simply wouldn't add the decorators to the corresponding mast nodes.
For `BasicBlock`s we currently store the `DecoratorList` alongside the basic block data, but since it is just referencing ops in the block and decorators, (it's a `Vec<(usize, DecoratorId)>`) we could move it to the decorator section and turn it into a `(MastNodeId, Vec<(usize, DecoratorId)>)`, so that it would work conceptually the same as the other decorators for deserialization: If that list is missing, we simply don't add it to the basic block.

I think that would be the optimal layout as it would allow us to remove all decrator data. Taking your comments into account also, the layout would then be:

```rust
//! The serialization format of MastForest is as follows:
//!
//! (Metadata)
//! - MAGIC
//! - VERSION
//!
//! (sections metadata)
//! - nodes length (`usize`)
//! - decorator data section offset (`usize`)
//! - decorators length (`usize`)
//!
//! (procedure roots section)
//! - procedure roots (`Vec<MastNodeId>`)
//!
//! (basic block data section)
//! - basic block data (previously called "node data", but I think it is more precise now)
//!
//! (node info section)
//! - MAST node infos (`Vec<MastNodeInfo>`)
//!
//! (decorator data section)
//! - Decorator data
//! - String table
//!
//! (decorator info section)
//! - decorator infos (`Vec<DecoratorInfo>`)
//!
//! (basic block decorator lists section)
//! - basic block decorator lists (`(MastNodeId, Vec<(usize, DecoratorId)>)`)
//! 
//! (before enter and after exit decorators section)
//! - before enter decorators (`Vec<(MastNodeId, Vec<DecoratorId>)>`)
//! - after exit decorators (`Vec<(MastNodeId, Vec<DecoratorId>)>`)
```

@yasonk So for this PR we would have to change:
- Add decorator data section offset to the header. Contrary to what I said before this is actually possible because we serialize the various data into `Vec<u8>`s first and then write those to the target, so we can inspect it for its length and add it to the header before writing those data.
- Serialize `DecoratorList` of `BasicBlocks` similarly to before-enter/after-exit decorators.

I think we also should add a `MastForest::truncate_decorators(serialized_mast_forest: &mut Vec<u8>) -> Result<(), MastForestError>` function which validates the MAGIC and VERSION and attempts to read the sections metadata. If it is all valid it drops the decorator data from the end and updates the `decorator data section offset` and `decorators length`.

Finally, we should add tests as mentioned before.

Reply:
@PhilippGackstatter Do you think it may be helpful to split this work into 2 PRs?
1. Place all decorator data at the end. (Good catch with the decorator list inside basic blocks)
2. Truncate decorators.

For (1), I would leave the current tests that make sure the data is serialized and deserialized correctly. Also include a fix for extracting decorator lists that are inside `node_data`.

For (2), we would implement the `truncate_decorators` along with adding state/logic needed to make that happen. E.g. updating `decorator data section offset` and `decorators_length`.
This would obviously require new tests to check that the decorators are properly removed.

Reply:
> Do you think it may be helpful to split this work into 2 PRs?

Whatever you prefer, I think both works. The `truncate_decorators` function shouldn't be a very large addition. Adding some tests for it and making sure that deserilization still works fine would be more code to add I would guess. Doing 2 PRs is also fine. After (1) the code should also be in a consistent state overall, that's the important part.

Btw re-reading the layout again I proposed, the `basic block decorator lists` should be a vector - I edited my comment.